---
title: "RNA-Seq Project Report"
author: '@tfogler'
date: "`r Sys.Date()`"
output: html_document
---

# Report

```{r setup, include=FALSE, results='hide'}
knitr::opts_chunk$set(echo = TRUE)
source('main.R')
```

## Discussion

**Remarks on the quality of sequencing reads:**

- Reads are high quality in general
  * No samples or reads need to be removed
- Only problems are the base sequence content and sequence duplication levels
  * sequence duplication is more worrisome since it could mean enrichment bias
  * on avg %s of reads after deduplication is 25%
  * a majority >50% also are duplicated 10x or more
  * base sequence content varies most at beginning of read
- After samtools flagstats, all reads for each sample passed QC


## Methods

First, mRNAseq samples were obtained corresponding to the wild-type/control (CTL) and knockout (KO) conditions, 3 CTL and 3 KO, and 2 reads. To ensure data quality, we performed initial quality control using FastQC v0.12.1.
Next, we aligned the reads to the gencode human primary assembly genome (GRCh38, release 45) using HISAT2 v2.2.1 with paired-end alignment using default parameters.
From these alignments, we generated paired-end gene (exon) counts using VERSE v0.1.5 and the gencode 45 primary assembly GTF.
After combining the counts from all 6 samples, we filtered out genes with zero counts. Then, we performed normalization and differential expression analysis using DESeq2 v1.44.0, comparing the two conditions.
Finally, we performed a functional gene set enrichment analysis using clusterProfiling v4.10.1 and org.Hs.eg.db Database to find functions
associated with differentially expressed genes, and a GO analysis of our significant gene list with DAVID.

Snakemake Workflow Summary
```{r DAG}
knitr::include_graphics("workflow_snake.png")
```

<!-- ![]("workflow_snake.png") -->

## Results
### R-side Analysis


```{r DESEQ}
#Read in verse_counts.tsv and display the head of the resulting tibble
counts_df <- load_n_trim("results/verse_counts.tsv", row.names="gene")
head(counts_df)

coldata <- data.frame(condition = rep(c("CTL", "KO"), each=3),
                      rep = rep(c("rep1", "rep2", "rep3"), times=2))
row.names(coldata) <- c("CTL_rep1", "CTL_rep2", "CTL_rep3", "KO_rep1", "KO_rep2", "KO_rep3")
cat("coldata: ")
coldata

# a warning about factors here is normal
dds <- run_deseq(counts_df, coldata, 10, "condition_CTL_vs_KO")
deseq_res <- results(dds)
cat("deseq results:")
head(deseq_res)

write.csv(deseq_res, file = "DESeq2_results.csv", quote = FALSE)
```
```{r histogram}
# get log 2 fold change from DESEQ results
log2FC <- deseq_res$log2FoldChange

hist(log2FC, main="Distribution of log2FoldChange of DE genes", ylab = "Frequency of genes")
```

<!-- pca -->
```{r pca plot}
#Create and display PCA plot
pca_plot <- plot_pca(dds, "DESeq Counts PCA")
pca_plot
```

This PCA plot shows the between-sample distance for the first two PCs. In it, the control replications are clustered fairly close together compared to the knockout replications. Two of the knockouts are very close to each other, but not the third, suggesting the third rep may be an outlier. 


```{r volcano plot}
# Post-DE Analysis

# FDR threshold: <0.05
fdr_threshold <- 0.05 
# Subset DE results based on FDR threshold
deseq_res <- na.omit(deseq_res)
significant_genes <- deseq_res[deseq_res[['padj']] < fdr_threshold,]
# View the number of significant genes
num_significant_genes <- nrow(significant_genes)
print(paste("Number of significant genes at FDR <", fdr_threshold, ":", num_significant_genes))

# Further analyze or visualize the significant genes as needed

# GENERATE VOLCANO PLOT
deseq_df <- as.data.frame(deseq_res)
volcano <- volcano_plot(deseq_df, "log2FoldChange", "padj", alpha=0.05)
ggsave("volcano1.svg", device='svg')
volcano
```

```{r GSEA analysis}
deseq_res <- deseq_res %>% 
  na.omit()

ranking <- "stat"

deseq_res <- deseq_res[order(-deseq_res[[ranking]]),]
head(deseq_res)
geneList <- deseq_res[[ranking]]
names(geneList) <- str_extract(rownames(deseq_res), "^\\w*")
gse <- clusterProfiler::gseGO(
  geneList = geneList,
  ont = "BP",
  keyType = "ENSEMBL",
  OrgDb = "org.Hs.eg.db",
  eps = 1e-10
)

# List results of GSEA as a beautiful Table
as.data.frame(gse)[1:3,] %>% knitr::kable()
write.csv(gse, "GSE_res.csv", quote=F)
```

```{r GS plot n list}
# Plot the 1st most significant GS pathway
clusterProfiler::gseaplot(gse, geneSetID = 1)

# Get list of most expressed genes as characters
top10_genes <- deseq_df[base::order(deseq_df[['padj']]), ][1:10, ] %>% rownames() %>% str_extract("^\\w+")
cat(top10_genes, sep = ', ')
```

```{r DAVID enrichment analsys}

# clusterProfiler::enrichDAVID(gene = top10_genes,
#                             idType="ENSEMBL_GENE_ID")

# read in results from david website
david_res <- read.delim("DAVID.csv")
as.data.frame(david_res) %>% knitr::kable()
```
### 4. How similar are the results of both analyses? Are there notable differences?

The results are shockingly different. DAVID GO analysis gave us genesets that have to do with plasma membrane, cytosol, and translation, 
whereas the most significant pathways in GSEA are to do with ATP synthesis in the mitochondrion and organ development being downregulated and pathways
to do with synaptic function being upregulated. Interestingly, you see a  more consistent pattern from the GSEA results. 